crates/bevy_ecs/src/storage/thin_array_ptr.rs

```diff
@@ -0,0 +1,250 @@
+    ///
+    /// # Safety
+    /// - `index` must be in bounds (`index` < `len`)
+    pub unsafe fn get_unchecked_raw(&mut self, index: usize) -> *mut T {
+        // SAFETY:
+        // - `self.data` and the resulting pointer are in the same allocated object
+        // - the memory adress of the last element doesn't overflow `isize`, so if `index` is in bounds, it won't overflow either
+        unsafe { self.data.as_ptr().add(index) }
+    }
+
+    /// Get a reference to the element at `index`. This method doesn't do any bounds checking.
+    ///
+    /// # Safety
+    /// - `index` must be in bounds (`index` < `len`)
+    /// - The element at index `index` must be safe to read
+    pub unsafe fn get_unchecked(&self, index: usize) -> &'_ T {
+        // SAFETY:
+        // - `self.data` and the resulting pointer are in the same allocated object
+        // - the memory adress of the last element doesn't overflow `isize`, so if `index` is in bounds, it won't overflow either
+        let ptr = unsafe { self.data.as_ptr().add(index) };
+
+        // SAFETY:
+        // - The pointer is properly aligned
+        // - It is derefrancable (all in the same allocation)
+        // - `index` < `len` and the element is safe to write to, so its valid
+        // - We have a reference to self, so no other mutable accesses to the element can occur
+        unsafe {
+            ptr.as_ref()
+                // SAFETY: We can use `unwarp_unchecked` because the pointer isn't null)
+                .unwrap_unchecked()
+        }
+    }
+
+    /// Get a mutable reference to the element at `index`. This method doesn't do any bounds checking.
+    ///
+    /// # Safety
+    /// - `index` must be in bounds (`index` < `len`)
+    /// - The element at index `index` must be safe to write to
+    pub unsafe fn get_unchecked_mut(&mut self, index: usize) -> &'_ mut T {
+        // SAFETY:
+        // - `self.data` and the resulting pointer are in the same allocated object
+        // - the memory adress of the last element doesn't overflow `isize`, so if `index` is in bounds, it won't overflow either
+        let ptr = unsafe { self.data.as_ptr().add(index) };
+
+        // SAFETY:
+        // - The pointer is properly aligned
+        // - It is derefrancable (all in the same allocation)
+        // - `index` < `len` and the element is safe to write to, so its valid
+        // - We have a mutable reference to `self`
+        unsafe {
+            ptr.as_mut()
+                // SAFETY: We can use `unwarp_unchecked` because the pointer isn't null)
+                .unwrap_unchecked()
+        }
+    }
+
+    // TODO: Docs
+    /// # Safety
+    /// The caller must:
+    /// - ensure that `index < len`
+    /// - ensure that `last_element_index` = `len - 1`
+    /// - update their saved length value to reflect that the last element has been removed (decrement it)
+    pub unsafe fn swap_remove_and_forget_unchecked(
```

Comment:
Unlike BlobVec, where we cannot load onto the stack, this type has type information while compiling, and you can just move the value onto the stack and return it directly. You can use `copy_nonoverlapping` instead of swap. See `RawVec::swap_remove`'s implementation in the standard library.